src/steps/csp.js

```diff
@@ -0,0 +1,116 @@
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+import crypto from 'crypto';
+import { select, selectAll } from 'hast-util-select';
+import { remove } from 'unist-util-remove';
+
+export const NONCE_AEM = '\'nonce-aem\'';
+
+function parseCSP(csp) {
+  const parts = csp.split(';');
+  const result = {};
+  parts.forEach((part) => {
+    const [directive, ...values] = part.trim().split(' ');
+    result[directive] = values.join(' ');
+  });
+  return result;
+}
+
+function shouldApplyNonce(csp) {
+  const parsedCSP = parseCSP(csp);
+  return {
+    scriptNonce: parsedCSP['script-src']?.includes(NONCE_AEM),
+    styleNonce: parsedCSP['style-src']?.includes(NONCE_AEM),
+  };
+}
+
+function createAndApplyNonce(res, tree, metaCSP, headersCSP) {
+  const nonce = crypto.randomBytes(16).toString('base64');
+  let scriptNonceResult = false;
+  let styleNonceResult = false;
+
+  if (metaCSP) {
+    const { scriptNonce, styleNonce } = shouldApplyNonce(metaCSP.properties.content);
+    scriptNonceResult ||= scriptNonce;
+    styleNonceResult ||= styleNonce;
+    metaCSP.properties.content = metaCSP.properties.content.replaceAll(NONCE_AEM, `'nonce-${nonce}'`);
+  }
+
+  if (headersCSP) {
+    const { scriptNonce, styleNonce } = shouldApplyNonce(headersCSP);
+    scriptNonceResult ||= scriptNonce;
+    styleNonceResult ||= styleNonce;
+    res.headers.set('content-security-policy', headersCSP.replaceAll(NONCE_AEM, `'nonce-${nonce}'`));
+  }
+
+  if (scriptNonceResult) {
+    selectAll('script[nonce="aem"]', tree).forEach((el) => {
+      el.properties.nonce = nonce;
+    });
+  }
+
+  if (styleNonceResult) {
+    selectAll('style[nonce="aem"]', tree).forEach((el) => {
+      el.properties.nonce = nonce;
+    });
+    selectAll('link[rel=stylesheet][nonce="aem"]', tree).forEach((el) => {
+      el.properties.nonce = nonce;
+    });
+  }
+}
+
+export function checkResponseBodyForMetaBasedCSP(res) {
+  return res.body?.includes('http-equiv="content-security-policy"')
+    || res.body?.includes('http-equiv="Content-Security-Policy"');
+}
+
+export function checkResponseBodyForAEMNonce(res) {
+  return res.body?.includes(NONCE_AEM);
+}
+
+export function getMetaCSP(tree) {
+  return select('meta[http-equiv="content-security-policy"]', tree)
+    || select('meta[http-equiv="Content-Security-Policy"]', tree);
+}
+
+export function getHeaderCSP(res) {
+  return res.headers?.get('content-security-policy');
+}
+
+export function contentSecurityPolicy(res, tree) {
+  const metaCSP = getMetaCSP(tree);
+  const headersCSP = getHeaderCSP(res);
+
+  if (!metaCSP && !headersCSP) {
+    // No CSP defined
+    return;
+  }
+
+  // CSP with nonce
+  if (
+    (metaCSP && metaCSP.properties.content.includes(NONCE_AEM))
+    || (headersCSP && headersCSP.includes(NONCE_AEM))
```

Comment:
would it be possible that a header uses double quotes?

Reply:
No, it is not possible. I tried it out both as meta tag and as header resulting the browser will show the following error  and a blank page, because no script loads.

```
The source list for the Content Security Policy directive 'script-src' contains an invalid source: '"nonce-r4Nd0mmm"'. It will be ignored.
```